content/blog/2020-09-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,85 @@
++++
+title = "Efficient Path Profiling"
+[extra]
+latex = true
+[[extra.authors]]
+name = "Theodore Bauer"
+bio = """
+  [Theodore Bauer][http://tedbauer.github.io/] is a first year M.S. student at Cornell who is interested in programming languages and systems. He likes spending time outside and making Spotify playlists.
+"""
+link = "http://tedbauer.github.io/"
++++
+
+Profiling a program provides important data that can guide program tuning, test coverage, compilation, and more. Researchers and engineers analyze various program aspects, like basic blocks and CFG edges. In particular, these two aspects are viewed as low-overhead although somewhat inaccurate with respect to path execution. Path profiling, which is at the level of sequences of basic block executions, can provide an accurate path perspective.
+
+[Our reading today][pathprofiling] provides a novel algorithm for profiling path executions with tolerable program overhead, making a compelling case for using path profiling.
+
+## The Algorithm
+
+This algorithm instruments programs in such a way that, at the end of a program execution, we are left with generated integer numbers that we can use to infer the paths that executed. At a high level, the algorithm does this for a program by:
+
+1. Assigning its CFG edges special weights
+2. Optimizing this weight selection to reduce overhead
+3. Adding instrumentation to its code that uses these weights to generate the path-identifying numbers.
+
+Additionally, the paper gives us an algorithm that lets us recover paths from the generated integers. Now, we'll dive into how each of these parts of the algorithm work.
+
+### 1. Assigning CFG edges special weights
+
+The algorithm starts by assigning every edge in the program's CFG a weight in such a way that, the sum of the weights across every path through the CFG maps to a unique number (no two paths map to the same number) that's at least $0$ and less than the number of paths.
```

Comment:
The comma in "that, the" is not necessary.

Reply:
Removed!